src/boxy.nim

```diff
@@ -361,121 +343,187 @@ proc grow(boxy: Boxy) =
-    return index
+  # Flush the draw calls to copy all images
+  boxy.flush()
 
-  boxy.grow()
-  boxy.takeFreeTile()
+  # Restore framebuffer binding
+  if boxy.layerNum >= 0:
+    glBindFramebuffer(GL_FRAMEBUFFER, boxy.layerFramebufferId)
+  else:
+    glBindFramebuffer(GL_FRAMEBUFFER, 0)
 
-proc removeImage*(boxy: Boxy, key: string) =
-  ## Removes an image, does nothing if the image has not been added.
-  if key in boxy.entriesBuffered:
-    raise newException(
-      BoxyError,
-      "Attempting to remove an image that is set to be drawn"
-    )
+  # Clean up the temporary framebuffer
+  glDeleteFramebuffers(1, growFramebufferId.addr)
 
-  if key in boxy.entries:
-    for tileLevel in boxy.entries[key].tiles:
-      for tile in tileLevel:
-        if tile.kind == tkIndex:
-          boxy.takenTiles.unsafeSetFalse(tile.index)
-    boxy.entries.del(key)
+  # Swap to the new atlas and entries
+  boxy.atlasTexture = newAtlasTexture
+  boxy.atlasSize = newAtlasSize
+  boxy.allocator = newAllocator
+  boxy.entries = newEntries
+
+  # Restore matrices
+  boxy.proj = savedProj
+  boxy.mat = savedMat
+  boxy.mats = savedMats
+
+  # Delete the old atlas texture
+  glDeleteTextures(1, oldAtlasTexture.textureId.addr)
 
-proc addImage*(boxy: Boxy, key: string, image: Image, genMipmaps = true) =
+proc addImage*(boxy: Boxy, key: string, image: Image) =
   if key in boxy.entriesBuffered:
     raise newException(
       BoxyError,
       "Attempting to modify an image that is already set to be drawn " &
       "(try using a unique key?)"
     )
 
-  boxy.removeImage(key)
-
-  boxy.entriesBuffered.incl(key)
+  # If image is one color:
+  if image.isOneColor():
+    var imageInfo = ImageInfo()
+    imageInfo.size = ivec2(image.width.int32, image.height.int32)
+    imageInfo.isOneColor = true
+    imageInfo.oneColor = image[0, 0].color
+    imageInfo.atlasPos = ivec2(0, 0)
+    boxy.entries[key] = imageInfo
+    return
 
+  # Check if the image is already in the atlas
   var imageInfo: ImageInfo
-  imageInfo.size = ivec2(image.width.int32, image.height.int32)
+  var reusing = false
+  if key in boxy.entries:
+    imageInfo = boxy.entries[key]
+    if imageInfo.cap.x >= image.width and imageInfo.cap.y >= image.height:
+      reusing = true
+
+  if not reusing:
+    # New image can't fit in the existing image info.
+    # Remove the existing image info and create a new one.
+    boxy.removeImage(key)
+    imageInfo = ImageInfo()
+    boxy.entriesBuffered.incl(key)
+    imageInfo.size = ivec2(image.width.int32, image.height.int32)
+    imageInfo.cap = imageInfo.size
+    imageInfo.isOneColor = false
+
+    # Try to pack the image using the allocator
+    var packed = false
+    var x, y: int
+    while not packed:
+      let allocation = boxy.allocator.allocate(image.width, image.height)
+      if allocation.success:
+        x = allocation.x
+        y = allocation.y
+        packed = true
+      else:
+        # Need to grow the atlas
+        boxy.grow()
```

Comment:
Could we pass a hint to grow to tell it how much free space in one rect it needs?
That way we avoid a degenerate case of multiple grows if it takes a few steps to reach the target size

Reply:
Yeah that could also be mitigated by having an addImages path, where they're sorted by size reverse (pack biggest first)

Reply:
"gc" then is at runtime when modifying the atlas more than from fragmenting the space on load

Reply:
Yeah totally.